DOCSP-39687 Create A Mongo Client (#39)
mongodb/docs-java-rs: source/connect-to-mongo/create-a-mongo-client.txt

.. _java-rs-mongo-client:

====================
Create a MongoClient
====================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: connection string, URI, server, settings, client

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

Overview
--------

To connect to a MongoDB deployment, you need two things:

- A **connection URI**, also known as a *connection string*, which tells the {+driver-short+}
  which MongoDB deployment to connect to.

> - A **MongoClient** instance, which creates the connection to the MongoDB deployment

Reviewer: ```suggestion
- A **MongoClient** instance, which creates connections to the MongoDB deployment
```

Reviewer: ✅ Resolved.

        .. TODO: uncomment out when java-rs-auth page made
        .. For more information about the ``authSource`` connection option, see
        .. :ref:`java-rs-auth`. 

   * - ``host[:port]``

     - Required. The host and port number where MongoDB is running. If you don't
       include a port number, the driver uses the default port, ``27017``.

   * - ``/defaultauthdb``
     
     - Optional. The authentication database to use if the
       connection string includes ``username:password@``
       authentication credentials but not the ``authSource`` option. If you don't include 
       this component, the client authenticates the user against the ``admin`` database.

   * - ``?<options>``

     - Optional. A query string that specifies connection-specific
       options as ``<name>=<value>`` pairs. 
        ..    TODO: uncomment out when connection options page made
        ..    See :ref:`java-rs-connection-options` for a full description of
        ..    these options.

For more information about creating a connection string, see
:manual:`Connection Strings </reference/connection-string>` in the
MongoDB Server documentation.

MongoClient
-----------

> To create a connection to MongoDB, pass a connection URI as a string to the
> ``MongoClient`` constructor. In the 

Reviewer: ```suggestion
To create a connection to MongoDB, pass a connection URI as a string to the
method constructing a ``MongoClient``. In the 
```

The method `MongoClients.create` is not a "constructor". "constructor" has a specific different meaning in Java.

Reviewer: Weirdly, I cannot resolve the conversations I started, so I will be marking them with

✅ Resolved.

following example, the driver uses a sample connection URI to connect to a MongoDB 
instance on port ``27017`` of ``localhost``: 

.. code-block:: java

    import com.mongodb.reactivestreams.client.MongoClients;
    import com.mongodb.reactivestreams.client.MongoClient;

    public class MongoConnection {
        public static void main(String[] args) {
            String uri = "mongodb://localhost:27017/";

>             MongoClient client = MongoClients.create(uri);

Reviewer: ```suggestion
            try (MongoClient client = MongoClients.create(uri)) {
                // use `client` here
            }
```

Let's write code such that it closes the created client. Otherwise, we encourage users to not close the clients, which potentially delays the release on server resources associated with the client.

Reviewer: The new code contains `client.close();` inside the `try`-with-resources statement. We should not call `client.close` ourselves, as the `try`-with-resources statement does that for us automatically. Note that the code I proposed does not have `client.close();`.

Reviewer: Ah, `client.close();` was removed in a later commit.

✅ Resolved.

        }
    }

.. tip:: Reusing Your Client

>    Because each ``MongoClient`` instance represents a pool of connections to the

Reviewer: ```suggestion
   Because each ``MongoClient`` instance pools connections to the
```

I don't think that `MongoClient` "represents" a pool of connections. It maintains (often multiple) connection pools, but it also maintains other information and resources, for example, a client-side view of the deployment it is connected to, thread pools (one would unlikely say that `MongoClient` represents a pool of threads).

Reviewer: ✅ Resolved.

   database, most applications require only a single instance of
   ``MongoClient``, even across multiple requests. 
.. TODO: comment out when java-rs-faq page is created
.. To learn more about connection pools, see the :ref:`FAQ <java-rs-faq>` page.

API Documentation
-----------------

To learn more about creating a ``MongoClient`` instance in the {+driver-short+},
see the following API documentation:

> - `MongoClient <https://mongodb.github.io/mongo-java-driver/5.1/apidocs/mongodb-driver-reactivestreams/com/mongodb/reactivestreams/client/MongoClient.html>`__ 

Reviewer: ```suggestion
- `MongoClient <https://mongodb.github.io/mongo-java-driver/5.1/apidocs/mongodb-driver-reactivestreams/com/mongodb/reactivestreams/client/MongoClients.html>`__ 
```

We should link to `MongoClients` (note `s` at the end), because the `MongoClient` interface documents nothing about creating objects of this type.

Reviewer: ✅ Resolved.

.. _java-rs-mongo-client:

====================
Create a MongoClient
====================

.. facet::
   :name: genre
   :values: reference

.. meta::

>    :keywords: connection string, URI, server, Atlas, settings, client

Reviewer 2: There's no mention of Atlas on this page, so probably good to remove here.

```suggestion
   :keywords: connection string, URI, server, settings, client
```

.. contents:: On this page
   :local:
   :backlinks: none

>    :depth: 2

Reviewer 2: There are no nested headings on this page so depth of 1 is fine here
```suggestion
   :depth: 1
```

       this component, the client authenticates the user against the ``admin`` database.

   * - ``?<options>``

     - Optional. A query string that specifies connection-specific
       options as ``<name>=<value>`` pairs. 
        ..    TODO: uncomment out when connection options page made
        ..    See :ref:`java-rs-connection-options` for a full description of
        ..    these options.

For more information about creating a connection string, see
:manual:`Connection Strings </reference/connection-string>` in the
MongoDB Server documentation.

MongoClient
-----------

To create a connection to MongoDB, pass a connection URI as a string to the
``MongoClient`` constructor. In the 
following example, the driver uses a sample connection URI to connect to a MongoDB 
instance on port ``27017`` of ``localhost``: 

.. code-block:: java

    import com.mongodb.reactivestreams.client.MongoClients;
    import com.mongodb.reactivestreams.client.MongoClient;

    public class MongoConnection {
        public static void main(String[] args) {
            String uri = "mongodb://localhost:27017/";

>             MongoClient Client = MongoClients.create(uri);

Reviewer 2: I think "client" should be lowercase
```suggestion
            MongoClient client = MongoClients.create(uri);
```

        }
    }

.. tip:: Reusing Your Client

   Because each ``MongoClient`` instance represents a pool of connections to the
   database, most applications require only a single instance of

>    ``MongoClient``, even across multiple requests. However, if you fork
>    a process, the child process *does* need its own ``MongoClient`` instance.
> .. TODO: comment out when java-rs-faq page is created
> .. To learn more, see the :ref:`FAQ <java-rs-faq>` page.

Reviewer 2: I think the forking process mentioned in this is specifically referring to a Python forking process. We can probably delete it here. Also I think instead of linking to the FAQ page, we should have a link to the connection pool page (when it's created). [Here's an example](https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/connection/connect/#mongoclient) in the Java driver docs
```suggestion
   ``MongoClient``, even across multiple requests.
.. TODO: comment out when java-rs-faq page is created
.. To learn more, see the :ref:`FAQ <java-rs-faq>` page.
```

Author: Jordan and I had Slack convo about this comment and decided to keep the link to the faq page, but just make it more clear that the faq page has a section about connection pools.
